src/interface/file.rs

```diff
@@ -122,168 +127,186 @@ impl EmulatedFile {
+        unsafe {
+            let mut fobj = self.fobj.lock();
+            let map = fobj.take().unwrap();
+
+            let (oldmap_addr, _oldlen, _cap) = map.into_raw_parts();
+            munmap(oldmap_addr as *mut c_void, self.mapsize);
+        }
+
+
+        let realfobj = self.realfobj.lock();
+        realfobj.set_len(self.filesize as u64)?;
+
         Ok(())
     }
 
+    fn remap_file(&mut self) {
+        let emfile: Vec<u8>;
+
+        self.mapsize = ((self.filesize / MAP_1GB) + 1) * MAP_1GB;
+
+        let realfobj = self.realfobj.lock();
+        let _lenres = realfobj.set_len(self.mapsize as u64);
+
+        unsafe {
+            let mut fobj = self.fobj.lock();
+            let map = fobj.take().unwrap();
+
+            let (oldmap_addr, oldlen, _cap) = map.into_raw_parts();
+            let newmap_addr = mremap(oldmap_addr as *mut c_void, oldlen, self.mapsize, MREMAP_MAYMOVE);
+            let _syncret = msync(newmap_addr, self.mapsize, MS_ASYNC);
+
+            emfile =  Vec::<u8>::from_raw_parts(newmap_addr as *mut u8, self.mapsize, self.mapsize);
+        }
+
+        self.fobj = Arc::new(Mutex::new(Some(emfile)));
+    }
+
     pub fn shrink(&mut self, length: usize) -> std::io::Result<()> {
 
         if length > self.filesize { 
             panic!("Something is wrong. {} is already smaller than length.", self.filename);
         }
-        match &self.fobj {
-            None => panic!("{} is already closed.", self.filename),
-            Some(f) => { 
-                let fobj = f.lock();
-                fobj.set_len(length as u64)?;
-                self.filesize = length;         
-                Ok(())
-            }
-        }
+        let realfobj = self.realfobj.lock();
+        // realfobj.set_len(length as u64)?;
+        self.filesize = length;   
+        drop(realfobj);
+        if self.filesize > self.mapsize { self.remap_file() }
+        Ok(())
     }
 
     // Read from file into provided C-buffer
     pub fn readat(&self, ptr: *mut u8, length: usize, offset: usize) -> std::io::Result<usize> {
+        
+        let mut real_length = length;
+        if offset + length > self.filesize { real_length = self.filesize - offset }
+
         let buf = unsafe {
             assert!(!ptr.is_null());
-            slice::from_raw_parts_mut(ptr, length)
+            slice::from_raw_parts_mut(ptr, real_length)
         };
 
-        match &self.fobj {
-            None => panic!("{} is already closed.", self.filename),
-            Some(f) => { 
-                let mut fobj = f.lock();
-                if offset > self.filesize {
-                  panic!("Seek offset extends past the EOF!");
-                }
-                fobj.seek(SeekFrom::Start(offset as u64))?;
-                let bytes_read = fobj.read(buf)?;
-                Ok(bytes_read)
-            }
+        let mut fobjopt = self.fobj.lock();
+        let fobj = fobjopt.as_mut().unwrap();
+        if offset > self.filesize {
```

Comment:
I feel like this should be at the start so we don't do all of the other code above and then check if offset is too much.

Reply:
and is it something that we have to panic over?

Reply:
true @yizhuoliang this should be an error not a panic